## Applied Data Science - Capstone Project

### Opening a Sushi Restaurant in Buenos Aires 

#### Import required libraries

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
from pandas.io.json import json_normalize
import requests
from bs4 import BeautifulSoup
!pip install geocoder
import geocoder
import os
!pip install folium
import folium # map rendering library
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
# Matplotlib and associated plotting modules
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors
%matplotlib inline

from sklearn.cluster import KMeans

print('Libraries imported.')

Libraries imported.


#### Take the data from the Wikipedia page into a DataFrame

In [2]:
res = requests.get("https://en.wikipedia.org/wiki/Neighbourhoods_of_Buenos_Aires")
soup = BeautifulSoup(res.content, 'lxml')
table = soup.find('table', class_= 'wikitable sortable')

In [3]:
bsas_df = pd.read_html(str(table))[0]
bsas_df.head(10)

Name  Area in km²  Population  Commune
0   Agronomía          2.1       13963       15
1     Almagro          4.1      128206        5
2   Balvanera          4.4      137521        3
3    Barracas          7.6       73377        4
4    Belgrano          6.8      126816       13
5       Boedo          2.6       45563        5
6   Caballito          6.8      170309        6
7   Chacarita          3.1       25778       15
8     Coghlan          1.3       18021       12
9  Colegiales          2.3       52391       13

Drop unnecessary columns

In [4]:
bsas_df.drop(['Area in km²', 'Commune'], axis=1, inplace = True)
bsas_df

Name  Population
0             Agronomía       13963
1               Almagro      128206
2             Balvanera      137521
3              Barracas       73377
4              Belgrano      126816
5                 Boedo       45563
6             Caballito      170309
7             Chacarita       25778
8               Coghlan       18021
9            Colegiales       52391
10         Constitución       41894
11               Flores      142695
12             Floresta       37247
13              La Boca       43413
14          La Paternal       19058
15              Liniers       42083
16            Mataderos       62206
17            Monserrat       39175
18         Monte Castro       32782
19        Nueva Pompeya       60465
20                Núñez       49019
21              Palermo      225245
22    Parque Avellaneda       51678
23     Parque Chacabuco       54638
24          Parque Chas       18926
25     Parque Patricios       37791
26        Puerto Madero         406
27             Recoleta      165494
28               Retiro       38635
29             Saavedra       48956
30        San Cristóbal       46494
31          San Nicolás       28667
32            San Telmo       23198
33      Vélez Sársfield       34084
34            Versalles       13556
35         Villa Crespo       83646
36     Villa del Parque       55502
37         Villa Devoto       67712
38  Villa General Mitre       34204
39         Villa Lugano      108170
40           Villa Luro       31859
41        Villa Ortúzar       21256
42     Villa Pueyrredón       38558
43           Villa Real       13681
44      Villa Riachuelo       13995
45     Villa Santa Rita       32248
46        Villa Soldati       39477
47        Villa Urquiza       85587

In [5]:
bsas_df.shape

(48, 2)

Rename columns: Name to Neighborhood

In [6]:
bsas_df.rename(columns = {'Name': 'Neighborhood'}, inplace=True)
bsas_df

Neighborhood  Population
0             Agronomía       13963
1               Almagro      128206
2             Balvanera      137521
3              Barracas       73377
4              Belgrano      126816
5                 Boedo       45563
6             Caballito      170309
7             Chacarita       25778
8               Coghlan       18021
9            Colegiales       52391
10         Constitución       41894
11               Flores      142695
12             Floresta       37247
13              La Boca       43413
14          La Paternal       19058
15              Liniers       42083
16            Mataderos       62206
17            Monserrat       39175
18         Monte Castro       32782
19        Nueva Pompeya       60465
20                Núñez       49019
21              Palermo      225245
22    Parque Avellaneda       51678
23     Parque Chacabuco       54638
24          Parque Chas       18926
25     Parque Patricios       37791
26        Puerto Madero         406
27             Recoleta      165494
28               Retiro       38635
29             Saavedra       48956
30        San Cristóbal       46494
31          San Nicolás       28667
32            San Telmo       23198
33      Vélez Sársfield       34084
34            Versalles       13556
35         Villa Crespo       83646
36     Villa del Parque       55502
37         Villa Devoto       67712
38  Villa General Mitre       34204
39         Villa Lugano      108170
40           Villa Luro       31859
41        Villa Ortúzar       21256
42     Villa Pueyrredón       38558
43           Villa Real       13681
44      Villa Riachuelo       13995
45     Villa Santa Rita       32248
46        Villa Soldati       39477
47        Villa Urquiza       85587

#### Get the geographical coordinates 

In [7]:
def get_latilong(neighborhood):
    lati_long_coords = None
    while(lati_long_coords is None):
        g = geocoder.arcgis('{}, Buenos Aires, Buenos Aires'.format(neighborhood))
        lati_long_coords = g.latlng
    return lati_long_coords

In [8]:
neighborhoods = bsas_df['Neighborhood']
coords = [get_latilong(neighborhood) for neighborhood in bsas_df["Neighborhood"].tolist()]

In [9]:
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])
bsas_df['Latitude'] = df_coords['Latitude']
bsas_df['Longitude'] = df_coords['Longitude']

In [10]:
bsas_df

Neighborhood  Population   Latitude  Longitude
0             Agronomía       13963 -34.592430 -58.496590
1               Almagro      128206 -34.611080 -58.430280
2             Balvanera      137521 -34.610110 -58.406020
3              Barracas       73377 -34.649900 -58.389100
4              Belgrano      126816 -34.561530 -58.457020
5                 Boedo       45563 -34.632280 -58.417790
6             Caballito      170309 -34.622180 -58.428580
7             Chacarita       25778 -34.583510 -58.452870
8               Coghlan       18021 -34.561580 -58.474280
9            Colegiales       52391 -34.575020 -58.447770
10         Constitución       41894 -34.626950 -58.382950
11               Flores      142695 -36.017465 -59.079428
12             Floresta       37247 -34.628980 -58.481590
13              La Boca       43413 -34.634540 -58.359390
14          La Paternal       19058 -34.606230 -58.475560
15              Liniers       42083 -34.641294 -58.539573
16            Mataderos       62206 -34.998888 -58.377706
17            Monserrat       39175 -34.616730 -58.387130
18         Monte Castro       32782 -34.619530 -58.505350
19        Nueva Pompeya       60465 -34.653290 -58.420820
20                Núñez       49019 -34.550830 -58.466560
21              Palermo      225245 -34.588450 -58.423430
22    Parque Avellaneda       51678 -34.644180 -58.483330
23     Parque Chacabuco       54638 -34.634160 -58.432960
24          Parque Chas       18926 -34.584130 -58.478140
25     Parque Patricios       37791 -34.636310 -58.401320
26        Puerto Madero         406 -34.608740 -58.362930
27             Recoleta      165494 -34.587540 -58.397170
28               Retiro       38635 -34.594750 -58.382730
29             Saavedra       48956 -37.603588 -62.394781
30        San Cristóbal       46494 -34.622880 -58.405320
31          San Nicolás       28667 -34.608180 -58.377570
32            San Telmo       23198 -34.616630 -58.370450
33      Vélez Sársfield       34084 -34.628170 -58.489660
34            Versalles       13556 -34.641294 -58.539573
35         Villa Crespo       83646 -34.596690 -58.442250
36     Villa del Parque       55502 -34.602440 -58.491010
37         Villa Devoto       67712 -34.588780 -58.508500
38  Villa General Mitre       34204 -34.611850 -58.471390
39         Villa Lugano      108170 -34.672640 -58.475460
40           Villa Luro       31859 -34.637450 -58.499660
41        Villa Ortúzar       21256 -34.581070 -58.469540
42     Villa Pueyrredón       38558 -34.584980 -58.503370
43           Villa Real       13681 -34.625440 -58.526770
44      Villa Riachuelo       13995 -34.690750 -58.471150
45     Villa Santa Rita       32248 -34.613920 -58.479440
46        Villa Soldati       39477 -34.660810 -58.441210
47        Villa Urquiza       85587 -34.569520 -58.489740

Save to csv file

In [11]:
bsas_df.to_csv('bsas_df.csv', index=False)

#### Create a map of Buenos AIres

In [12]:
address = 'Buenos Aires, Argentina'

geolocator = Nominatim(user_agent="BsAs")
location = geolocator.geocode(address)
latitude = location.latitude
longitude= location.longitude
print('The geograpical coordinate of Buenos Aires are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Buenos Aires are -34.6075682, -58.4370894.


In [13]:
map_bsas = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, Neighborhood in zip(bsas_df['Latitude'], bsas_df['Longitude'], bsas_df['Neighborhood']):
    label = '{}'.format(Neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_bsas)  
    
map_bsas

#### Use Foursquare to explore the neighborhoods

In [14]:
CLIENT_ID = 'MLUPTXHJ1Q5H5F5S2ND4LIWBFD2S2AD3EP2HGSTNECXSG1CX' # my Foursquare ID
CLIENT_SECRET = '0ZAWTZ4V0YIYSBQMWB503XT4DUBLOSUMMZ3SUBO53QWKESD5' # my Foursquare Secret
VERSION = '20180604'

print('Your credentails:')
print('CLIENT_ID: '+CLIENT_ID)
print('CLIENT_SECRET: '+CLIENT_SECRET)

Your credentails:
CLIENT_ID: MLUPTXHJ1Q5H5F5S2ND4LIWBFD2S2AD3EP2HGSTNECXSG1CX
CLIENT_SECRET: 0ZAWTZ4V0YIYSBQMWB503XT4DUBLOSUMMZ3SUBO53QWKESD5


In [15]:
radius = 500
limit = 100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list =[]
    for name, lat, lng, in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
   
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venues_list in venues_list for item in venues_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
bsas_venues = getNearbyVenues(names= bsas_df['Neighborhood'],
                                 latitudes=bsas_df['Latitude'],
                                 longitudes=bsas_df['Longitude']
                                )

In [17]:
print(bsas_venues.shape)
bsas_venues.head(n=10)

(1114, 7)


Neighborhood  Latitude  Longitude                        Venue  \
0    Agronomía -34.59243  -58.49659                     Bonafide   
1    Agronomía -34.59243  -58.49659                       Baraba   
2    Agronomía -34.59243  -58.49659                 Bien de Bien   
3    Agronomía -34.59243  -58.49659      Renatto Cucina Italiana   
4    Agronomía -34.59243  -58.49659       Plaza Martín Rodríguez   
5    Agronomía -34.59243  -58.49659  Ladobueno Patisserie & Café   
6    Agronomía -34.59243  -58.49659     Repostería Papá Eduvilio   
7    Agronomía -34.59243  -58.49659                Club El Talar   
8    Agronomía -34.59243  -58.49659                    Rancho IN   
9    Agronomía -34.59243  -58.49659                    Al Piatto   

   Venue Latitude  Venue Longitude      Venue Category  
0      -34.590722       -58.498184         Coffee Shop  
1      -34.590674       -58.500094          Restaurant  
2      -34.592322       -58.500959                Café  
3      -34.591342       -58.500781  Italian Restaurant  
4      -34.590837       -58.501098               Plaza  
5      -34.596536       -58.498617         Coffee Shop  
6      -34.595906       -58.499338              Bakery  
7      -34.591820       -58.498330         Sports Club  
8      -34.595990       -58.495916          Steakhouse  
9      -34.590089       -58.497244         Pizza Place

Let's check how many venues are on each neighborhood

In [18]:
bsas_venues.groupby('Neighborhood').count()

Latitude  Longitude  Venue  Venue Latitude  \
Neighborhood                                                      
Agronomía                  18         18     18              18   
Almagro                    29         29     29              29   
Balvanera                  22         22     22              22   
Barracas                    4          4      4               4   
Belgrano                   42         42     42              42   
Boedo                       9          9      9               9   
Caballito                  41         41     41              41   
Chacarita                  45         45     45              45   
Coghlan                    23         23     23              23   
Colegiales                 27         27     27              27   
Constitución               18         18     18              18   
Flores                      2          2      2               2   
Floresta                    9          9      9               9   
La Boca                    12         12     12              12   
La Paternal                 7          7      7               7   
Liniers                     8          8      8               8   
Monserrat                  32         32     32              32   
Monte Castro                6          6      6               6   
Nueva Pompeya               7          7      7               7   
Núñez                      38         38     38              38   
Palermo                    70         70     70              70   
Parque Avellaneda           5          5      5               5   
Parque Chacabuco           11         11     11              11   
Parque Chas                11         11     11              11   
Parque Patricios           19         19     19              19   
Puerto Madero              36         36     36              36   
Recoleta                   59         59     59              59   
Retiro                    100        100    100             100   
San Cristóbal              28         28     28              28   
San Nicolás                54         54     54              54   
San Telmo                  94         94     94              94   
Versalles                   8          8      8               8   
Villa Crespo               66         66     66              66   
Villa Devoto               15         15     15              15   
Villa General Mitre        11         11     11              11   
Villa Lugano                9          9      9               9   
Villa Luro                 15         15     15              15   
Villa Ortúzar              23         23     23              23   
Villa Pueyrredón            7          7      7               7   
Villa Real                  7          7      7               7   
Villa Riachuelo             3          3      3               3   
Villa Santa Rita            7          7      7               7   
Villa Soldati               3          3      3               3   
Villa Urquiza              23         23     23              23   
Villa del Parque           22         22     22              22   
Vélez Sársfield             9          9      9               9   

                     Venue Longitude  Venue Category  
Neighborhood                                          
Agronomía                         18              18  
Almagro                           29              29  
Balvanera                         22              22  
Barracas                           4               4  
Belgrano                          42              42  
Boedo                              9               9  
Caballito                         41              41  
Chacarita                         45              45  
Coghlan                           23              23  
Colegiales                        27              27  
Constitución                      18              18  
Flores                             2               2  
Floresta                          

Let's find out the unique categories

In [19]:
print('There are {} unique categories.'.format(len(bsas_venues['Venue Category'].unique())))

There are 176 unique categories.


In [20]:
bsas_venues['Venue Category'].unique()[:50]

array(['Coffee Shop', 'Restaurant', 'Café', 'Italian Restaurant', 'Plaza',
       'Bakery', 'Sports Club', 'Steakhouse', 'Pizza Place', 'Bar',
       'Soccer Field', 'Train Station', 'Ice Cream Shop',
       'Argentinian Restaurant', 'Bus Station', 'Hotel',
       'Fast Food Restaurant', 'Empanada Restaurant',
       'Bike Rental / Bike Share', 'Concert Hall', 'Shop & Service',
       'Pharmacy', 'Dessert Shop', 'Gym', 'Pub', 'Grocery Store',
       'Cultural Center', 'Chinese Restaurant', 'Peruvian Restaurant',
       'BBQ Joint', 'Gas Station', 'Electronics Store', 'Bus Stop',
       'Platform', 'Convenience Store', 'Park', 'Farmers Market',
       'Athletics & Sports', 'Tea Room', 'Market', 'Salad Place',
       'Art Museum', 'Bookstore', 'French Restaurant', 'Smoke Shop',
       'Vegetarian / Vegan Restaurant', 'Comic Shop', 'German Restaurant',
       'Bike Shop', 'History Museum'], dtype=object)

#### Let's analyze each neighborhood

In [21]:
# one hot encoding
bsas_onehot = pd.get_dummies(bsas_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
bsas_onehot['Neighborhood'] = bsas_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [bsas_onehot.columns[-1]] + list(bsas_onehot.columns[:-1])
bsas_onehot = bsas_onehot[fixed_columns]

bsas_onehot.head()

Neighborhood  Accessories Store  American Restaurant  \
0    Agronomía                  0                    0   
1    Agronomía                  0                    0   
2    Agronomía                  0                    0   
3    Agronomía                  0                    0   
4    Agronomía                  0                    0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Bagel Shop  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bakery  Bar  Basketball Court  Bed & Breakfast  Beer Bar  Beer Garden  \
0       0    0                 0                0         0            0   
1       0    0                 0                0         0            0   
2       0    0                 0                0         0            0   
3       0    0                 0                0         0            0   
4       0    0                 0                0         0            0   

   Big Box Store  Bike Rental / Bike Share  Bike Shop  Bistro  Bookstore  \
0              0                         0          0       0          0   
1              0                         0          0       0          0   
2              0                         0          0       0          0   
3              0                         0          0       0          0   
4              0                         0          0       0          0   

   Breakfast Spot  Brewery  Buffet  Burger Joint  Bus Station  Bus Stop  Café  \
0               0        0       0             0            0         0     0   
1               0        0       0             0            0         0     0   
2               0        0       0             0            0         0     1   
3               0        0       0             0            0         0     0   
4               0        0       0             0            0         0     0   

   Camera Store  Candy Store  Chinese Restaurant  Chocolate Shop  Church  \
0             0            0                   0               0       0   
1             0            0                   0               0       0   
2             0            0                   0               0       0   
3             0            0                   0               0       0   
4             0            0                   0               0       0   

   Clothing Store  Cocktail Bar  Coffee Shop  Comfort Food Restaurant  \
0               0             0            1                        0   
1               0             0            0                        0   
2               0             0            0                        0   
3               0             0            0                        0   
4               0             0            0                        0   

   Comic Shop  Concert Hall  Construction & Landscaping  Convenience Store  \
0           0             0                           0                  0   
1           0             0                           0                  0   
2           0             0                           0                  0   
3           0             0                           0                  0   
4           0             0 

In [22]:
bsas_onehot.shape

(1114, 177)

Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [23]:
bsas_grouped = bsas_onehot.groupby('Neighborhood').mean().reset_index()
bsas_grouped

Neighborhood  Accessories Store  American Restaurant  \
0             Agronomía           0.000000             0.000000   
1               Almagro           0.000000             0.000000   
2             Balvanera           0.000000             0.000000   
3              Barracas           0.000000             0.000000   
4              Belgrano           0.000000             0.000000   
5                 Boedo           0.000000             0.000000   
6             Caballito           0.000000             0.000000   
7             Chacarita           0.000000             0.000000   
8               Coghlan           0.000000             0.000000   
9            Colegiales           0.000000             0.000000   
10         Constitución           0.000000             0.000000   
11               Flores           0.000000             0.000000   
12             Floresta           0.000000             0.000000   
13              La Boca           0.000000             0.000000   
14          La Paternal           0.000000             0.000000   
15              Liniers           0.000000             0.000000   
16            Monserrat           0.000000             0.000000   
17         Monte Castro           0.000000             0.000000   
18        Nueva Pompeya           0.000000             0.000000   
19                Núñez           0.000000             0.000000   
20              Palermo           0.042857             0.000000   
21    Parque Avellaneda           0.000000             0.000000   
22     Parque Chacabuco           0.000000             0.000000   
23          Parque Chas           0.000000             0.000000   
24     Parque Patricios           0.000000             0.000000   
25        Puerto Madero           0.000000             0.027778   
26             Recoleta           0.000000             0.016949   
27               Retiro           0.000000             0.000000   
28        San Cristóbal           0.000000             0.000000   
29          San Nicolás           0.000000             0.000000   
30            San Telmo           0.000000             0.000000   
31            Versalles           0.000000             0.000000   
32         Villa Crespo           0.000000             0.000000   
33         Villa Devoto           0.000000             0.000000   
34  Villa General Mitre           0.000000             0.000000   
35         Villa Lugano           0.000000             0.000000   
36           Villa Luro           0.000000             0.000000   
37        Villa Ortúzar           0.000000             0.000000   
38     Villa Pueyrredón           0.000000             0.000000   
39           Villa Real           0.000000             0.000000   
40      Villa Riachuelo           0.000000             0.000000   
41     Villa Santa Rita           0.000000             0.000000   
42        Villa Soldati           0.000000             0.000000   
43        Villa Urquiza           0.000000             0.000000   
44     Villa del Parque           0.000000             0.000000   
45      Vélez Sársfield           0.000000             0.000000   

    Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0.055556     0.000000    0.000000             0.000000   
1                 0.034483     0.000000    0.000000             0.000000   
2                 0.000000     0.000000    0.000000             0.000000   
3                 0.000000     0.000000    0.000000             0.000000   
4                 0.071429     0.000000    0.023810             0.000000   
5                 0.555556     0.000000    0.000000             0.000000   
6                 0.048780     0.000000    0.000000             0.000000   
7                 0.044444     0.000000    0.000000             0.000000   
8                 0.086957     0.000000    0.000000             0.000000   
9                 0.074074     0.000000    0.000000             0.000000   
10                0.00000

In [24]:
bsas_grouped.shape

(46, 177)

Let's print each neighborhood that has Sushi Restaurants

In [25]:
len(bsas_grouped[bsas_grouped['Sushi Restaurant'] >0])
bsas_sushi = bsas_grouped[['Neighborhood', 'Sushi Restaurant']]
bsas_sushi.head()

Neighborhood  Sushi Restaurant
0    Agronomía           0.00000
1      Almagro           0.00000
2    Balvanera           0.00000
3     Barracas           0.00000
4     Belgrano           0.02381

#### Cluster Neighborhoods

In [26]:
# Using K-Means to cluster neighborhood into  clusters
bsas_clustering = bsas_sushi.drop('Neighborhood', 1)
kmeans = KMeans(n_clusters=4, random_state=0).fit(bsas_clustering)
kmeans.labels_

array([0, 0, 0, 0, 3, 0, 1, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0,
       0, 0, 0, 0, 0, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0], dtype=int32)

Let's create a new dataframe the includes the cluster as well as the top 10 venues for each neighborhood

In [27]:
bsas_merged = bsas_sushi.copy()
bsas_merged["Cluster Labels"] = kmeans.labels_
bsas_merged.head()

Neighborhood  Sushi Restaurant  Cluster Labels
0    Agronomía           0.00000               0
1      Almagro           0.00000               0
2    Balvanera           0.00000               0
3     Barracas           0.00000               0
4     Belgrano           0.02381               3

In [28]:
# merge bsas_grouped with bsas_df to add latitude/longitude for each neighborhood
bsas_merged = bsas_merged.join(bsas_df.set_index('Neighborhood'), on='Neighborhood')
bsas_merged.head()

Neighborhood  Sushi Restaurant  Cluster Labels  Population  Latitude  \
0    Agronomía           0.00000               0       13963 -34.59243   
1      Almagro           0.00000               0      128206 -34.61108   
2    Balvanera           0.00000               0      137521 -34.61011   
3     Barracas           0.00000               0       73377 -34.64990   
4     Belgrano           0.02381               3      126816 -34.56153   

   Longitude  
0  -58.49659  
1  -58.43028  
2  -58.40602  
3  -58.38910  
4  -58.45702

Let's visualize the clusters

In [29]:
kclusters = 4

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi , cluster in zip(bsas_merged['Latitude'], 
                                   bsas_merged['Longitude'], 
                                   bsas_merged['Neighborhood'], 
                                   bsas_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' -Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [30]:
map_clusters.save('map_clusters.html')

#### Examine the clusters

In [31]:
bsas_merged.loc[bsas_merged['Cluster Labels']==0]

Neighborhood  Sushi Restaurant  Cluster Labels  Population  \
0             Agronomía               0.0               0       13963   
1               Almagro               0.0               0      128206   
2             Balvanera               0.0               0      137521   
3              Barracas               0.0               0       73377   
5                 Boedo               0.0               0       45563   
7             Chacarita               0.0               0       25778   
8               Coghlan               0.0               0       18021   
10         Constitución               0.0               0       41894   
11               Flores               0.0               0      142695   
12             Floresta               0.0               0       37247   
13              La Boca               0.0               0       43413   
14          La Paternal               0.0               0       19058   
15              Liniers               0.0               0       42083   
16            Monserrat               0.0               0       39175   
17         Monte Castro               0.0               0       32782   
18        Nueva Pompeya               0.0               0       60465   
19                Núñez               0.0               0       49019   
21    Parque Avellaneda               0.0               0       51678   
22     Parque Chacabuco               0.0               0       54638   
23          Parque Chas               0.0               0       18926   
24     Parque Patricios               0.0               0       37791   
25        Puerto Madero               0.0               0         406   
26             Recoleta               0.0               0      165494   
29          San Nicolás               0.0               0       28667   
30            San Telmo               0.0               0       23198   
31            Versalles               0.0               0       13556   
32         Villa Crespo               0.0               0       83646   
33         Villa Devoto               0.0               0       67712   
34  Villa General Mitre               0.0               0       34204   
35         Villa Lugano               0.0               0      108170   
36           Villa Luro               0.0               0       31859   
37        Villa Ortúzar               0.0               0       21256   
38     Villa Pueyrredón               0.0               0       38558   
39           Villa Real               0.0               0       13681   
40      Villa Riachuelo               0.0               0       13995   
41     Villa Santa Rita               0.0               0       32248   
42        Villa Soldati               0.0               0       39477   
43        Villa Urquiza               0.0               0       85587   
44     Villa del Parque               0.0               0       55502   
45      Vélez Sársfield               0.0               0       34084   

     Latitude  Longitude  
0  -34.592430 -58.496590  
1  -34.611080 -58.430280  
2  -34.610110 -58.406020  
3  -34.649900 -58.389100  
5  -34.632280 -58.417790  
7  -34.583510 -58.452870  
8  -34.561580 -58.474280  
10 -34.626950 -58.382950  
11 -36.017465 -59.079428  
12 -34.628980 -58.481590  
13 -34.634540 -58.359390  
14 -34.606230 -58.475560  
15 -34.641294 -58.539573  
16 -34.616730 -58.387130  
17 -34.619530 -58.505350  
18 -34.653290 -58.420820  
19 -34.550830 -58.466560  
21 -34.644180 -58.483330  
22 -34.634160 -58.432960  
23 -34.584130 -58.478140  
24 -34.636310 -58.401320  
25 -34.608740 -58.362930  
26 -34.587540 -58.397170  
29 -34.608180 -58.377570  
30 -34.616630 -58.370450  
31 -34.641294 -58.539573  
32 -34.596690 -58.442250  
33 -34.588780 -58.508500  
34 -34.611850 -58.471390  
35 -34.672640 -58.475460  
36 -34.637450 -58.499660  
37 -34.581070 -58.469540  
38 -34.584980 -58.503370  
39 -34.625440 -58.526770  
40 -34.690750 -58.471150  
41 -34.613920 -58.479440  
42 -34.660810 -58.

In [32]:
bsas_merged.loc[bsas_merged['Cluster Labels']==1]

Neighborhood  Sushi Restaurant  Cluster Labels  Population  Latitude  \
6    Caballito           0.04878               1      170309 -34.62218   

   Longitude  
6  -58.42858

In [33]:
bsas_merged.loc[bsas_merged['Cluster Labels']==2]

Neighborhood  Sushi Restaurant  Cluster Labels  Population  Latitude  \
9      Colegiales          0.037037               2       52391 -34.57502   
28  San Cristóbal          0.035714               2       46494 -34.62288   

    Longitude  
9   -58.44777  
28  -58.40532

In [34]:
bsas_merged.loc[bsas_merged['Cluster Labels']==3]

Neighborhood  Sushi Restaurant  Cluster Labels  Population  Latitude  \
4      Belgrano          0.023810               3      126816 -34.56153   
20      Palermo          0.014286               3      225245 -34.58845   
27       Retiro          0.010000               3       38635 -34.59475   

    Longitude  
4   -58.45702  
20  -58.42343  
27  -58.38273

#### Observations:

As you can see from analyzing the clusters, the sushi restaurants are located in the clusters 1, 2 and 3, leaving cluster 0 with no sushi restaurants at all. This gives us a great oportunity for these neighborhoods. We can then analyze by looking at the population of each neighborhood in cluster 0, and choose the one where the possibility of opening a sushi restaurant would be excellent. 

### Conclusion:
The neighborhood of Recoleta, with a population of 165494 inhabitants, would be a very good option for opening a sushi restaurant.
It is one of the most populated neighborhoods in cluster 0, and, as you can see from the map, it is located near the port, which means the number of passing by people is very high. I believe this is a good location for opening a sushi restaurant.